### ISSUE TITLE
android_camerax plugin returns .temp extension for recorded video

### ISSUE THREAD

**Issue by ekuleshov** [flutter/flutter#148335]
### Steps to reproduce

1. Launch example app from the camerax module
2. Start video recording
3. Stop video recording and observe extension of the reported saved file name.

### Expected results

The file returned from the stop video recording method should have a file extension corresponding to the saved file type. E.g. .mp4, .jpg etc.

### Actual results

Plugin saves recorded file with the .temp extension which does not correspond to the actual media type.

### Code sample

<details open><summary>Code sample</summary>

```dart
example app from the camerax module
```

</details>

### Screenshots or Video

N/A

### Logs

N/A

### Flutter Doctor output

N/A

**Comment by darshankawar** [comment]
Thanks for the report. Using https://github.com/flutter/packages/tree/main/packages/camera/camera_android_camerax/example and running on S10, I was able to observe that the recording link that gets saved with extension .temp.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.22.0, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.22.0 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5dcb86f68f (5 days ago), 2024-05-09 07:39:20 -0500
    • Engine revision f6344b75dc
    • Dart version 3.4.0
    • DevTools version 2.34.3
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.22.0-30.0.pre.11, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.22.0-30.0.pre.11 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c90e18c6f5 (2 hours ago), 2024-05-09 20:32:11 -0700
    • Engine revision c0917b14fc
    • Dart version 3.5.0 (build 3.5.0-138.0.dev)
    • DevTools version 2.36.0-dev.5
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by ekuleshov** [comment]
I should also say that this is a change of behavior from the old android camera plugin. And generally it is a breaking change. Please raise severity of this issue.

**Comment by wdcs-dimilkalathiya** [comment]
This issue is breaking our current implementation is there any update regarding this? we have temporarily downgraded plugin and it works for now.

I would like to know if there is workaround this issue while using android_camerax

**Comment by AlexV525** [comment]
cc @camsim99

**Comment by camsim99** [comment]
@gmackall Correct me if I'm wrong, but I believe we recently discussed this and determined we plan to leave this as is because we define in the plugin where to save the recorded video, and when we do that before the recording starts, we do not know the file type. According to https://developer.android.com/media/camera/camerax/video-capture#bind-videocapture,

> The `Recorder` selects the most suitable format for the system. The most common video codec is [H.264 AVC](https://developer.android.com/reference/android/media/MediaFormat#MIMETYPE_VIDEO_AVC)) with container format [MPEG-4](https://developer.android.com/reference/android/media/MediaFormat#MIMETYPE_VIDEO_MPEG4).

With this in mind, we could assume it is a mp4 like the example code in the same docs show. I'm not currently aware of a way to determine the type of recording before configuring the `Recorder`.

**Comment by ekuleshov** [comment]
@camsim99 the issue is about the file extension returned from the stop video recording method.

The `.temp` extension returned now is not useful for any practical purposes and would have to be changed. To not have every consumer of android_camerax plugin have to deal with that, it would be great if plugin returned a proper file extension that matches the recorded file content.

**Comment by gmackall** [comment]
Spoke more with @camsim99 offline - we define the output path before starting the video but that output path is only referenced internally until we return the file. It should be possible for us to determine the file type after recording has stopped, but before returning from [`stopVideoRecording`](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart#L996), so that we return a file with the correct extension.

**Comment by Ne0n11** [comment]
Can't the temp file for video be .mp4 format from the start? It is like that in camera_android version

**Comment by jimonik** [comment]
Which version did not return. temp format？

**Comment by wdcs-dimilkalathiya** [comment]
> Which version did not return. temp format？

use ```camera: 0.10.6```

**Comment by wzslr321** [comment]
> on returned now is not useful for any practical purposes and would have to be changed. To not have every consumer of android_camerax plugin have to deal with that, it would be great if plugin returned a proper file extension that matches the recorded file content.

+1

**Comment by thomas4Bitcraft** [comment]
Still an issue - really annoying as it changes the way the file must be handled in android only

**Comment by camsim99** [comment]
> Spoke more with @camsim99 offline - we define the output path before starting the video but that output path is only referenced internally until we return the file. It should be possible for us to determine the file type after recording has stopped, but before returning from [`stopVideoRecording`](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart#L996), so that we return a file with the correct extension.

Not actively working on this but happy to take patches to fix this!! This should be the general approach, but happy to provide further code pointers for anyone willing to take it on.

**Comment by ekuleshov** [comment]
> Not actively working on this but happy to take patches to fix this!! This should be the general approach, but happy to provide further code pointers for anyone willing to take it on.

Please provide these pointers. I could try to look at it, as this is one of the blocking issues preventing moving to the camerax plugin.

**Comment by camsim99** [comment]
> It should be possible for us to determine the file type after recording has stopped, but before returning from [`stopVideoRecording`](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart#L996), so that we return a file with the correct extension.

@ekuleshov Dug into it and I'm not 100% sure how that approach would work but I do see a world where we choose a common format and return that instead if there's a good argument for it. I say this because this is what [the documentation](https://developer.android.com/reference/androidx/camera/video/Recorder#prepareRecording(android.content.Context,androidx.camera.video.FileOutputOptions):~:text=A%20recorder%20can%20be%20used%20to%20save%20the%20video%20frames%20sent%20from%20the%20VideoCapture%20use%20case%20in%20common%20recording%20formats%20such%20as%20MPEG4.) seems to suggest. I don't know that we can determine the file type vs set it.

We prepare the file to put the recording in here: https://github.com/flutter/packages/blob/da4c3eedb5e70d47452088d8cb30c02c058052a7/packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/RecorderHostApiImpl.java#L101, so we could choose an mp4 file or something like that. Do you have thoughts on this?

**Comment by ekuleshov** [comment]
> @ekuleshov Dug into it and I'm not 100% sure how that approach would work but I do see a world where we choose a common format and return that instead if there's a good argument for it. I say this because this is what [the documentation](https://developer.android.com/reference/androidx/camera/video/Recorder#prepareRecording(android.content.Context,androidx.camera.video.FileOutputOptions):~:text=A%20recorder%20can%20be%20used%20to%20save%20the%20video%20frames%20sent%20from%20the%20VideoCapture%20use%20case%20in%20common%20recording%20formats%20such%20as%20MPEG4.) seems to suggest. I don't know that we can determine the file type vs set it.

The legacy `camera_android` plugin sets the `.mp4` file extension unconditionally and then returns it as is when recording is stopped:

https://github.com/flutter/packages/blob/main/packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java#L1238-L1244

> We prepare the file to put the recording in here: https://github.com/flutter/packages/blob/da4c3eedb5e70d47452088d8cb30c02c058052a7/packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/RecorderHostApiImpl.java#L101, so we could choose an mp4 file or something like that. Do you have thoughts on this?

It looks like it is either too late or there is not enough context about the file in that place.

However on the Flutter side, the `AndroidCameraCameraX` implementation of the `CameraPlatform` sets the `videoOutputPath` (with the `.temp` extension) in the `startVideoCapturing()` method and then simply returns a file at that path when the `stopVideoRecording()` is called.

https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart#L1083-L1086

Maybe @gmackall could chime in why the `.temp` extension was used unlike the `.mp4` in the legacy camera plugin.

I see two options:
* Just use the `.mp4` extension as in the old plugin
* Autodetect the file type and rename `.temp` file accordingly.

For the second option the [mime](https://pub.dev/packages/mime) package from the tools.dart.dev could be used. Perhaps something like this in the `stopVideoRecording()` method:

```
    final File videoOutputFile = File(videoOutputPath!);
    final List<int> header = videoOutputFile.openSync().readSync(defaultMagicNumbersMaxLength);
    final String? mimeType = lookupMimeType(videoOutputPath!, headerBytes: header);
    if (mimeType != null) {
      final String? extension = extensionFromMime(mimeType);
      if (extension != null) {
        videoOutputPath = videoOutputFile.renameSync(videoOutputPath!.replaceAll(RegExp(r'\.temp'), '.$extension')).path;
      }
    }
```

PS: Interestingly, there is an old issue https://github.com/flutter/flutter/issues/91680 requesting an option to allow to specify the output path and it looks like having the `videoOutputPath` in the camerax implementation would allow to implement it easily (that is if the Android code will have permissions to write to that external file path).

**Comment by camsim99** [comment]
@ekuleshov sorry for the late response and thank you so much for the thorough research!

I think it's reasonable to have the CameraX plugin match the implementation of the old plugin. Ideally, though, if we could autodetect the type and set it like in the way you suggested, then that's an improvement over camera_android. @gmackall @reidbaker any opinions here?

Also interesting point about that proposal. We would need to consider other platforms as well for that proposal, though.

**Comment by reidbaker** [comment]
https://developer.android.com/media/camera/camerax/video-capture#configure-and-create-recording

<img width="877" height="229" alt="Image" src="https://github.com/user-attachments/assets/7def4bf8-8ecb-4f3e-8014-41a966e8f1fe" />

This documentation makes it sound like there is no way to *guarantee* you can pick the format and also that if you did always assume mp4 then you would be right *most* of the time. Which sounds like a great way to author a difficult for someone to understand bug.

@camsim99 can you reach out to the camerax team and see if there is a way to know what format they chose (which is different than setting a format we want)?

**Comment by ekuleshov** [comment]
@reidbaker that's why I've mentioned a [2nd option in my comment above](https://github.com/flutter/flutter/issues/148335#issuecomment-2455924283) to verify the file type after recording is done and rename extension accordingly before returning file to consumer.

**Comment by reidbaker** [comment]
@ekuleshov do you know what apis we can use to determine video format? Because conceptually I understood your comment but that isnt enough for us to author the change. My hope was there was a way to specify format and that looks to not be possible the backup of finding an api where camerax can tell us the appropriate extension is a good fallback. Reading the file and running it through a decoder or analyzing file headers requires that we would need to know all possible file output types which I would like to avoid if possible.

**Comment by ekuleshov** [comment]
> [@ekuleshov](https://github.com/ekuleshov) do you know what apis we can use to determine video format? Because conceptually I understood your comment but that isnt enough for us to author the change. My hope was there was a way to specify format and that looks to not be possible the backup of finding an api where camerax can tell us the appropriate extension is a good fallback. Reading the file and running it through a decoder or analyzing file headers requires that we would need to know all possible file output types which I would like to avoid if possible.

You don't think [the code snippet linked above](https://github.com/flutter/flutter/issues/148335#issuecomment-2455924283) is not going to work? It uses Dart's `mime` package to detect video file type and corresponding extension and then renames video file in place on the Dart side. Or it is  not appropriate for other reasons?

The `mime` package only need to read a small bit of the video file to identify type and it already supports number of file types. I'd think that should cover all known video types.

There is probably a similar equivalent of mime type detection on the Android side of things too if that's more preferable.

**Comment by reidbaker** [comment]
@ekuleshov sorry I saw that on friday and not on my reread this morning. Yes that snippet is what I was hoping to see. I still think an api from android is better but that is a good alternative. 

Thank you for your engagement on this issue I really do appreciate it.

**Comment by ekuleshov** [comment]
> [@ekuleshov](https://github.com/ekuleshov) ... I still think an api from android is better but that is a good alternative.

@reidbaker there are some complications with Android API. The `ContentResolver` wants a `ContentResolver.SCHEME_CONTENT` scheme `Uri` instance, which may require a `FileProvider` configured in the app (unless I'm missing something):

```
File videoFile ...

Uri contentUri = FileProvider.getUriForFile(context, 
                context.getPackageName() + ".fileprovider", file);
        
String mimeType = context.getContentResolver().getType(contentUri);

String ext = android.webkit.MimeTypeMap.getSingleton()
    .getExtensionFromMimeType(mimeType)
```

**Comment by camsim99** [comment]
Followed up with the CameraX team about this. What I learned:

**TLDR:** It's probably safe to assume MP4.

- CameraX aims to produce MP4 files (this is confirmed by [the Android docs](https://developer.android.com/reference/androidx/camera/video/Recorder#:~:text=recorder%20can%20be%20used%20to%20save%20the%20video%20frames%20sent%20from%20the%20VideoCapture%20use%20case%20in%20common%20recording%20formats%20such%20as%20MPEG4.))
- Some legacy or low-end devices may use 3GPP instead. However, the video is often still playable when saved with an .mp4 extension
- The file format is derived from [CamcorderProfile.fileFormat](https://developer.android.com/reference/android/media/CamcorderProfile#fileFormat), so technically other formats besides MP4 are possible, but they've received no issues regarding incorrect file extensions (implying that assuming MP4 is ok)
- There is no CameraX API to check for the file extensions because they've received no relevant issues regarding incorrect file extensions 

Unless there's major pushback, I want to go ahead and have the plugin assume MP4 so we're unlikely to run into issues and will be more useful for developers than a .temp.

**Comment by ekuleshov** [comment]
I guess assuming mp4 may work in most cases... Though that isn't that far off from checking the actual file type and renaming a returned/completed file with the proper extension for that type.

The statement about video file saved with an incorrect .mp4 extension being playable may not hold true on the Flutter side of things and there are more things that could be done with that vide besides playing it.

**Comment by camsim99** [comment]
@ekuleshov Sure but aren't you saying the only way to do that is for "FileProvider [to be] configured in the app"?

> The statement about video file saved with an incorrect .mp4 extension being playable may not hold true on the Flutter side of things and there are more things that could be done with that vide besides playing it.

Also, can you elaborate on this? Why would this hold true for Android and not Flutter?

**Comment by ekuleshov** [comment]
> [@ekuleshov](https://github.com/ekuleshov) Sure but aren't you saying the only way to do that is for "FileProvider [to be] configured in the app"?

Probably, if its done in the Android code. But can be also done on the Dart side of things.

> > The statement about video file saved with an incorrect .mp4 extension being playable may not hold true on the Flutter side of things and there are more things that could be done with that vide besides playing it.
> 
> Also, can you elaborate on this? Why would this hold true for Android and not Flutter?

Flutter is multi-platform and potentially file captured on Android could be played or processed/decoded somewhere else (e.g. using 3rd party non-Android-only video decoders, e.g. vlc, ffmpeg, etc). Even Flutter plugins running on Android that aren't using ExoPlayer for playing videos may not be able to play files saved with an incorrect extension. Or they would have to run their own file type detection in order to be able to do correct video decoding.
